### PR TITLE
[Spark] Support Spark Structured Logging API

### PR DESCRIPTION
**Pull request by zedtang**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Uber issue: #3145 

Support Spark Structured Logging API by using the shimming layer. Spark Structured Logging is only available in Spark 4.0 snapshot so need to shim the API to make it compile for Spark 3.5.
## How was this patch tested?

Tests that the new API on Spark master can produce structured logs: DeltaStructuredLoggingSuite
Tests that the shimming API on Spark 3.5 still produce the same plain text logs: DeltaPatternLoggingSuite
## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala
Comment (gengliangwang) [inline spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala:47]: There is a recent changes in https://github.com/apache/spark/pull/46634/files#diff-41ea575e2e1305fa8a92f4e3f549e7b896d97eac0e43e00d3dc65c816846e349R52. Let's wait until the PR is merged (probably this week)

File: spark/src/main/scala/org/apache/spark/sql/delta/logging/LogKeys.scala
Comment (gengliangwang) [inline spark/src/main/scala/org/apache/spark/sql/delta/logging/LogKeys.scala:47]: QQ: I plan to reduce the number of log keys in Apache Spark. Meanwhile, there will be new log keys from delta in this file. How will we keep sync with the Apache Spark?
Shall we just store the log keys for delta logging in this file?

Comment (zedtang) [reply spark/src/main/scala/org/apache/spark/sql/delta/logging/LogKeys.scala:47]: Yeah we can and should just store the keys that Delta uses. This means potentially duplicate some keys with Spark in this file for the keys that Delta uses but that's fine.

I removed the used keys.

**Review by gengliangwang — APPROVED** [review]
LGTM

File: spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala
Comment (jaceklaskowski) [inline spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala:44]: Why not singular (`LogKeyShim`)?

File: spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala
Comment (jaceklaskowski) [inline spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala:24]: Since it's a new file, use 2024

File: spark/src/main/scala-spark-3.5/shims/LoggingShims.scala
Comment (jaceklaskowski) [inline spark/src/main/scala-spark-3.5/shims/LoggingShims.scala:46]: s/cannot/must

Please add the type to the error OR remove the error message altogether (since it does not add much anyway).

File: spark/src/main/scala-spark-3.5/shims/LoggingShims.scala
Comment (jaceklaskowski) [inline spark/src/main/scala-spark-3.5/shims/LoggingShims.scala:76]: A singular?

**Review by jaceklaskowski — CHANGES_REQUESTED** [review]

File: spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala
Comment (zedtang) [reply spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala:24]: It looks like every file under spark/ always uses 2021.

File: spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala
Comment (zedtang) [reply spark/src/main/scala-spark-3.5/shims/LogKeyShims.scala:44]: This follows the naming convention under https://github.com/delta-io/delta/tree/master/spark/src/main/scala-spark-3.5/shims

File: spark/src/main/scala-spark-3.5/shims/LoggingShims.scala
Comment (zedtang) [reply spark/src/main/scala-spark-3.5/shims/LoggingShims.scala:46]: This is a stripped-down version from https://github.com/apache/spark/blob/bdcb79f23da3d09469910508426a54a78adcbda6/common/utils/src/main/scala/org/apache/spark/internal/Logging.scala#L78 to provide the syntax support, without any functionality.
I prefer to keep it as is if it doesn't fix anything.

File: spark/src/main/scala-spark-3.5/shims/LoggingShims.scala
Comment (zedtang) [reply spark/src/main/scala-spark-3.5/shims/LoggingShims.scala:76]: ditto

**Review by zedtang — COMMENTED** [review]
@jaceklaskowski which of your comments are blocking? I see you have used the request changes option

**Pull request merged**
